lib/translate/inquiry.cc

```diff
@@ -0,0 +1,200 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "inquiry.h"
+#include "string.h"
+namespace inquiry {
+    // Creates and validates a Inquiry Command struct
+    std::optional<scsi_defs::InquiryCommand> raw_cmd_to_scsi_command(absl::Span<const uint32_t> raw_cmd) {
+        if (raw_cmd.empty()) {
+            printf("buffer is empty or nullptr\n");
+            return std::nullopt;
+        }
+
+        uint8_t opcode = raw_cmd[0];
+        if (static_cast<scsi_defs::OpCode>(opcode) != scsi_defs::OpCode::kInquiry) {
+            printf("invalid opcode. expected %ux. got %ux.", static_cast<uint8_t>(scsi_defs::OpCode::kInquiry), opcode);
+            return std::nullopt;
+        }
+
+        // printf("opcode is good, casting\n");
+        // TODO: check invalid parameters in scsi_command
+        return std::make_optional(*reinterpret_cast<const scsi_defs::InquiryCommand*>(&raw_cmd[1]));
+    }
+
+    // Executes the NVME Identify Controller command
+    nvme_defs::IdentifyControllerData nvme_identify_controller() {
```

Comment:
AFAIK lib/translate is intended to strictly act as a land for code to live which translates between SCSI and NVMe. A lot of functionality picked up in this file, including activities such as executing NVMe commands, renders the translation library reliant on downstream dependencies. The mini-engines are probably better placed in kernel module code at /modules. I attached a flow-chart to the Google doc which details an overview of how this code could be segmented

Reply:
Alternatively, we could create a lib/engine directory for c++ engine code to live

Reply:
Just to clarify, according to the flow-chart, the code inside lib/translate should just be responsible for 
1) translating SCSI -> NVMe commands
2) translating NVMe -> SCSI responses
3) 2 pass memory translation

Everything else, including the parsing from raw bytes to scsi commands would be doone in the mini engine, which exists as a kernel module? 

I also agree that the translation lib should have the single responsibility of translation

Reply:
The three core responsibilities of lib/translation that you outlined are the intention behind the flowchart, yeah. Just trying to keep different parts of the code organized with a clear and manageable scope.

There's still some unhandled cases which the chart doesn't explicitly deal with, like, as you mentioned, taking SCSI command buffers and casting them to our custom structures. I think writing a library for this task to be accessed from the engine/kernel module and thus enjoying the benefits of Bazel testing is a perfectly reasonable thing to do.

Things that should probably exist as inseparable from the engine are processes like memory allocation, as that is inherently tied to our full-scope implementation.

Reply:
That's a good idea-putting the modules that the engine deals with in other scopes / files. I will look into reorganizing.